### PR TITLE
Add `try_catch()`

### LINKED ISSUE #503: with_return() ?
```R
try_parse_eval <- function(text, env = globalenv()) {
  callCC(function(return) {
    expr <- tryCatch(parse(text = text), error = function(e) return(NULL))
    res <- tryCatch(eval(expr, env), error = function(e) return(NULL))
    res
  })
}

# ->

try_parse_eval <- function(text, env = globalenv()) {
  with_return({
    expr <- tryCatch(parse(text = text), error = function(e) return(NULL))
    res <- tryCatch(eval(expr, env), error = function(e) return(NULL))
  })
  res
}
```

### PR DESCRIPTION
**Pull request by lionel-**
Closes #503. Also deprecate (silently for now) `with_handlers()`.

`try_catch()` is about as fast as `tryCatch()` with 1 handler, and faster from 2 handlers on.

```r
cnd <- errorCondition("")
f <- function() signalCondition(cnd)

throw_identity <- function(cnd, throw) throw(cnd)

bench::mark(
  tryCatch = tryCatch(f(), error = identity),
  try_catch = try_catch(f(), error = throw_identity)
)[1:8]
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 tryCatch     8.92µs 10.3µs    81089.        0B     0    10000     0
#> 2 try_catch   10.12µs   11µs    87145.        0B     8.72  9999     1

bench::mark(
  tryCatch = tryCatch(f(), error = identity, warning = identity),
  try_catch = try_catch(f(), error = throw_identity, warning = throw_identity)
)[1:8]
#>   expression     min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 tryCatch    12.7µs  13.9µs    59361.        0B     5.94  9999     1
#> 2 try_catch     11µs  11.8µs    80371.        0B     0    10000     0
```

It creates leaner call stacks:

```r
f <- function() g()
g <- function() h()
h <- function() abort("foo")

tryCatch(f(), message = identity, warning = identity)
last_trace()
#> <error/rlang_error>
#> Error in `h()`: foo
#> Backtrace:
#>      ▆
#>   1. ├─base::tryCatch(f(), message = identity, warning = identity)
#>   2. │ └─base tryCatchList(expr, classes, parentenv, handlers)
#>   3. │   ├─base tryCatchOne(...)
#>   4. │   │ └─base doTryCatch(return(expr), name, parentenv, handler)
#>   5. │   └─base tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
#>   6. │     └─base tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>   7. │       └─base doTryCatch(return(expr), name, parentenv, handler)
#>   8. └─global f()
#>   9.   └─global g()
#>  10.     └─global h()

try_catch(f(), message = throw_identity, warning = throw_identity)
last_trace()
#> <error/rlang_error>
#> Error in `h()`: foo
#> Backtrace:
#>     ▆
#>  1. ├─rlang::try_catch(f(), message = throw_identity, warning = throw_identity)
#>  2. │ └─base::withCallingHandlers(...)
#>  3. └─global f()
#>  4.   └─global g()
#>  5.     └─global h()
```

### CODE REVIEW

File: R/cnd-handlers.R [hunk 1]
```diff
@@ -1,8 +1,109 @@
...
+#' as second argument.
+#'
+#' ```{r, error = TRUE}
+#' # A prototype of error handler that does not do anything
+#' try_catch(1 + "", error = function(cnd, throw) { NULL })
+#' ```
+#'
+#' A condition handler can:
+#'
+#' - Interrupt the computation of `expr` and return from `try_catch()`
```
[the remark this answers is not in the document]
Comment (lionel-) [reply R/cnd-handlers.R:18]: I initially tried that but the code blocks caused the bullet list to close in the generated Rd. Maybe that's ok?

[the remark this answers is not in the document]
Comment (hadley) [reply R/cnd-handlers.R:18]: Did you indent them four spaces? (you have to switch from 2 spaces to 4 spaces when the bullets themselves contain other components)

File: R/cnd-handlers.R [hunk 2]
```diff
@@ -1,8 +1,136 @@
...
+#'
+#' A condition handler can:
+#'
+#' -   *Recover from conditions** with a value. In this case the computation of
+#'     `expr` is aborted and the recovery value is returned from
+#'     `try_catch()`. Error recovery is useful when you don't want
+#'     errors to abruptly interrupt your program but resume at the
+#'     catching site instead.
+#'
+#'     ```
```
Comment (hadley) [inline R/cnd-handlers.R:19]: I tweaked these to not run, so that the code is correctly indented (this must be a roxygen2 bug). If you think the output is really important, you could just manually simulate it.

File: R/cnd-handlers.R [hunk 3]
```diff
@@ -1,8 +1,136 @@
-#' Establish handlers on the stack
+#' Try an expression with condition handlers
 #'
 #' @description
 #' `r lifecycle::badge("experimental")`
 #'
+#' `try_catch()` establishes handlers for conditions of a given class
+#' (`"error"` `"warning"`, `"message"`, ...). Handlers are functions
```
Comment (hadley) [inline R/cnd-handlers.R:7]:
```suggestion
#' (`"error"`, `"warning"`, `"message"`, ...). Handlers are functions
```

File: R/cnd-handlers.R [hunk 4]
```diff
@@ -1,8 +1,136 @@
...
+#'     add information to low-level errors about the high-level context
+#'     in which they occurred.
+#'
+#'     ```
+#'     try_catch(1 + "", error = function(cnd) abort("Failed.", parent = cnd))
+#'     ```
+#'
+#' -   **Inspect conditions**, for instance to log data about warnings
+#'     or errors. In this case, the handler must return the [zap()]
+#'     sentinel to instruct `tryCatch()` to ignore (or zap) that
```
Comment (hadley) [inline R/cnd-handlers.R:35]:
```suggestion
#'     sentinel to instruct `try_catch()` to ignore (or zap) that
```

File: R/cnd-handlers.R [hunk 5]
```diff
@@ -1,8 +1,136 @@
...
+#'     ```
+#'
+#' @param expr An R expression.
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Named condition
+#'   handlers. The names specify the condition class for which a
+#'   handler will be called.
+#'
+#'   Handler functions must have at least two arguments. By
+#'   convention, the second argument should be named `throw`.
+#'
```
Comment (hadley) [inline R/cnd-handlers.R:54]:
```suggestion
```

File: R/cnd-handlers.R [hunk 6]
```diff
@@ -1,8 +1,136 @@
...
+#' in a single function. It reproduces the behaviour of both calling
+#' and exiting handlers depending the on the return value of the
+#' handler. If the handler returns the [zap()] sentinel, it is taken
+#' as a calling handler that declines to recover from a condition.
+#' Otherwise, it is taken as an exiting handler which returns a value
+#' from the catching site.
+#'
+#' There is an important difference between `tryCatch()` and
+#' `try_catch()`. The latter doesn't catch condition objects, instead
+#' it catches recovery values returned by the handler. The main
```
Comment (hadley) [inline R/cnd-handlers.R:92]: I don't understand what this means. Could you reword it to say something like "preserves the call stack"?

File: R/cnd-handlers.R [hunk 7]
```diff
@@ -1,8 +1,136 @@
...
+#' consequence is that the program is still fully running when an
+#' error handler is called. This makes it possible to capture a full
+#' backtrace for the error, e.g. when rethrowing the error with
+#' `abort(parent = cnd)`. Technically, `try_catch()` is more similar
+#' to (and implemented on top of) [base::withCallingHandlers()] than
+#' `tryCatch().`
+#'
+#' A big downside to keeping the call stack alive when handlers are run
+#' is that stack overflow can't be handled with `try_catch()`.
+#'
```
Comment (hadley) [inline R/cnd-handlers.R:102]:
```suggestion
```

File: R/cnd-handlers.R [hunk 8]
```diff
@@ -1,8 +1,136 @@
...
+
+  throw <- function(x) {
+    value <<- x
+    catch
+  }
+
+  .External(ffi_try_catch, frame)
+}
+
+utils::globalVariables("catch")
```
Comment (hadley) [inline R/cnd-handlers.R:118]: I wonder if it's slightly nicer to do `value <- catch <- NULL` above?

File: src/internal/cnd-handlers.c [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  // Build arguments of the form `function(cnd) handlers[[1]](cnd, throw)`
+  r_obj* args = r_null;
+  r_keep_t shelter; KEEP_HERE(args, &shelter);
+
+  for (int i = n - 1; i >= 0; --i) {
+    r_obj* cls = v_classes[i];
+    r_obj* hnd = KEEP(r_copy(hnd_call));
+
+    r_obj* subscript_node = r_node_cdr(r_node_cdar(r_node_cadr(r_node_cdar(r_node_cdar(r_node_cddr(hnd))))));
```
Comment (hadley) [inline src/internal/cnd-handlers.c:27]: Worth a comment explaining what R code this equivalent to?

File: src/internal/cnd-handlers.c [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+#include "decl/cnd-handlers-decl.h"
+
+r_obj* ffi_try_catch(r_obj* try_catch_args) {
+  r_obj* env = r_node_cadr(try_catch_args);
+
+  r_obj* handlers = KEEP(rlang_env_dots_list(env));
+  r_env_poke(env, rlang_syms.handlers, handlers);
+
+  r_obj* classes = r_names(handlers);
+  if (classes == r_null) {
```
Comment (hadley) [inline src/internal/cnd-handlers.c:12]: This check causes `try_catch(1 + 1)` to error. I think I'd expect it to work.

**Review by hadley — APPROVED** [review]

File: src/internal/cnd-handlers.c (hunk 10 above)
Comment (lionel-) [reply src/internal/cnd-handlers.c:12]: Good catch, this is especially important since we support `!!!` and the handlers list size might vary.
